core/src/Cabana_MemberTypes.hpp

```diff
@@ -74,6 +74,54 @@ struct MemberTypeAtIndex<M, MemberTypes<Types...>>
     using type = typename MemberTypeAtIndexImpl<M, Types...>::type;
 };
 
+//---------------------------------------------------------------------------//
+/*!
+  \class CheckMemberTypes
+  \brief Check that member types are valied.
+*/
+template <std::size_t M, typename T, typename... Types>
+struct CheckMemberTypesImpl;
+
+template <typename T, typename... Types>
+struct CheckMemberTypesImpl<0, T, Types...>
+{
+    using type = T;
+    static_assert( std::is_trivial<type>::value,
+                   "Member types must be trivial" );
+
+    using value_type = typename std::remove_all_extents<type>::type;
+    static_assert( std::is_arithmetic<value_type>::value,
+                   "Member value types must be arithmetic" );
```

Comment:
Only built-in types will have a specialization of `is_arithmetic` with a value that is `true` and all are trivial as far as I know.